doc/programming_guide/asrc/software_architecture.rst

```diff
@@ -0,0 +1,179 @@
+
+**usb_to_i2s_intertile** task receives the ASRC output data generated by usb_audio_out_asrc over the inter-tile context onto the |I2S| tile and writes it to the |I2S| send_buffer. It has other rate-monitoring related responsibilities that are described in the rate server section.
+
+**i2s_to_usb_intertile** task receives the ASRC output data generated by i2s_audio_recv_asrc over the inter-tile context onto the USB tile and writes it to the USB samples_to_host_stream_buf. It has other rate-monitoring related responsibilities that are described in the rate server section.
+
+The diagram below shows the application tasks involved in the |I2S| → ASRC → USB path processing and their interaction with each other.
+
+.. figure:: diagrams/asrc_i2s_to_usb_data_path.png
+   :align: center
+   :scale: 80 %
+   :alt: ASRC |I2S| → ASRC → USB data path
+
+The diagram below shows the application tasks involved in the USB → ASRC → |I2S| path processing and their interaction with each other.
+
+.. figure:: diagrams/asrc_usb_to_i2s_data_path.png
+   :align: center
+   :scale: 80 %
+   :alt: ASRC |I2S| → ASRC → USB data path
+
+
+.. _rate-server-label:
+
+rate_server
+-----------
+The ASRC process_frame API requires the caller to calculate and send the instantaneous ratio between the ASRC input and output rate. The rate_server is responsible for calculating these rate ratios for both USB → ASRC → |I2S| and |I2S| → ASRC → USB directions.
+
+Additionally, the application also monitors the average buffer fill levels of the buffers holding ASRC output to prevent any overflows or underflows of the respective buffer. A gradual drift in the buffer fill level indicates that the rate-ratio is being under or over calculated by the rate_server. This could happen either due to jitter in the actual rates or precision limitations when calculating the rates.
+
+The average fill level of the buffer is monitored and a closed-loop error correction factor is calculated to keep the buffer level at an expected stable level. The error estimated based on the buffer fill level is used to compute the estimated rate ratio from the initial rate ratio. This estimated rate ratio is then sent to the ASRC process_frame() API.
+
+.. code-block:: console
+
+    estimated_rate_ratio = initial_rate_ratio + buffer_based_correction_factor
+
+The rate_server runs on the |I2S| tile (tile 1) and is periodically triggered from the USB tile (tile 0) by the usb_to_i2s_intertile task. The rate_server is triggered once after every 16 frames are written to the samples_to_host_stream_buf.
+
+The following information is needed for calculating the rate ratios
+
+1. Average |I2S| rate
+2. Average USB rate
+3. An error factor computed based on the USB samples_to_host_stream_buf fill level
+4. An error factor computed based on the |I2S| send buffer fill level
+5. USB mic interface open flag indicating if the EP1 IN interface is open since ASRC_I2S_to_USB_rate_ratio is calculated only if the mic interface is open
```

Comment:
I couldn't find either ASRC_I2S_to_USB_rate_ratio or  ASRC_USB_to_I2S_rate_ratio in the code. Are they correct?

Reply:
They're not. I got tired of writing the long form again and again. I've fixed it. Thanks for spotting!